### ISSUE TITLE
🐛(WinOS) `--config-path` causes "unexpected error"

### ISSUE THREAD

**Issue by rivy** [biomejs/biome#3479]
### Environment information

```block
CLI:
  Version:                      1.8.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           unset
  JS_RUNTIME_NAME:              unset
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### What happened?

`biome lint --config-path .\biome.jsonc` panics, where otherwise, `biome lint` completes correctly.

```shell
❯ biome lint --config-path .\biome.jsonc
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ignore-0.4.21\src\gitignore.rs:229:9
Thread Name: biome::worker_0
Message: path is expected to be under the root
```

### Expected result

It should not panic.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by Conaclos** [comment]
@ematipico Are you able to reproduce this bug?

**Comment by ematipico** [comment]
The error comes from the `ignore` crate, so it's possible we are passing an incorrect path. However, we need a reproduction. This is an issue generated by the VCS integration

**Comment by rivy** [comment]
@ematipico , why are you closing this issue? It's a fundamental problem when running the application.
😕

**Comment by ematipico** [comment]
We requested a reproduction, but that was never provided. After so much inactivity, we close the issue.

**Comment by rivy** [comment]
I don’t understand.
A reproduction was included in the initial post about the issue. It includes the simple WinOS command line which prompts the bug and the output (including the request to post it here to help you fix the issue).

**Comment by rivy** [comment]
@Conaclos , notably:

1 -  an empty or missing `.gitignore` avoids the error
2 -  a `.gitignore` with only comments avoids the error
3 - a `.gitignore` with any actual "ignore" (non-commentary) content => ERROR

I hope that's helpful.

**Comment by Conaclos** [comment]
Unfortunately I cannot reproduce the issue.

Here is my attempt:

```shell
❯ git init

❯ npm init --yes

❯ npm i -D @biomejs/biome

❯ tree 
├── biome.jsonc
├── package.json
├── src
│   └── index.js

❯ cat biome.jsonc
{
  "vcs": {
    "enabled": true,
    "clientKind": "git",
    "useIgnoreFile": true
  }
}

❯ npx biome lint --config-path ./biome.jsonc 
Checked 3 files in 909µs. No fixes applied.

```

Note: I am on Linux.

**Comment by rivy** [comment]
@Conaclos ,

Yes, it looks like a WinOS-specific problem.
Using WSL with the exact same project files, there is no error.

**Comment by rivy** [comment]
Here's a minimal reproduction (file repo at <https://github.com/rivy-fix/biome-config-bug>).

```shell
>  dir
 Volume in drive C is Local Disk
 Volume Serial Number is 1C33-48C1

 Directory of C:\Users\Roy\AARK\Projects\js\t-biome\repo

2024-08-29  20:22    <DIR>          .
2024-08-29  20:22    <DIR>          ..
2024-08-29  20:24    <DIR>          .git
2024-08-29  20:21                 3 .gitignore
2024-08-29  20:19               137 biome.json
2024-08-29  20:09               329 package.json
               3 File(s)            469 bytes
               3 Dir(s)  369,052,930,048 bytes free
> cat .gitignore
FOO
> cat biome.json
{
        "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
        "vcs": { "clientKind": "git", "enabled": true, "useIgnoreFile": true }
}
> cat package.json
{
  "name": "repo",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "Roy Ivy III <rivy.dev@gmail.com> (https://rivy.dev/)",
  "license": "MIT",
  "description": "",
  "devDependencies": {
    "@biomejs/biome": "^1.8.3"
  }
}
> npx biome lint
Checked 2 files in 1988µs. No fixes applied.
> npx biome lint --config-path biome.json
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ignore-0.4.21\src\gitignore.rs:229:9
Thread Name: biome::worker_0
Message: path is expected to be under the root
```

**Comment by ematipico** [comment]
Thank you. That's what I meant with providing a reproduction. Providing only the commands isn't enough.

**Comment by rivy** [comment]
Ping @ematipico , any progress on this?
It's still an issue as of v1.9.4.
Thanks for any updates.

**Comment by ematipico** [comment]
I'm sorry, but I don't own a WSL, so I can't reproduce the problem. If you're able to send a fix, that would be much appreciated. 

This where you should look at: https://github.com/biomejs/biome/blob/c9d0ba86ba62f4e3de66b12c26d98173be15eab2/crates/biome_service/src/configuration.rs#L172

**Comment by rivy** [comment]
@ematipico , I'll try to look at it when I have some time.

To be clear about the bug though... it's *not* a WSL issue. WSL (and POSIX-like) seem to work correctly. It's only *WinOS* (for CMD/PowerShell) that looks to be failing.

**Comment by ematipico** [comment]
You're right. It's an issue related to the VCS git integration and the config path.

I think I know what's going on. Keep the reproduction available, I might be able to look at it next week

**Comment by ematipico** [comment]
@rivy 

I found the problem. Our CLI doesn't accept Windows paths, only Unix paths. Can you try again and use `biome lint --config-path=./biome.json` instead?

**Comment by rivy** [comment]
> [@rivy](https://github.com/rivy)
> 
> I found the problem. Our CLI doesn't accept Windows paths, only Unix paths. Can you try again and use `biome lint --config-path=./biome.json` instead?

Nope. It looks like the app is tacking on `\.biome.json` onto whatever I pass as a file path.

```shell
❯ biome lint --config-path=./biome.json
./biome.json\biome.json internalError/fs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Cannot read file

  × Biome can't read the following file, maybe for permissions reasons or it doesn't exist: ./biome.json\biome.json

❯ biome lint --config-path=biome.json
./biome.json\biome.json internalError/fs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Cannot read file

  × Biome can't read the following file, maybe for permissions reasons or it doesn't exist: ./biome.json\biome.json

❯ biome lint --config-path=.
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ignore-0.4.21\src\gitignore.rs:229:9
Thread Name: biome::worker_5
Message: path is expected to be under the root

❯ biome lint --config-path=xxx
xxx\biome.json internalError/fs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Cannot read file

  × Biome can't read the following file, maybe for permissions reasons or it doesn't exist: xxx\biome.json
```

**Comment by rivy** [comment]
@empatico, I've traced the problem.

It's not a path problem per se. I've traced the `--config-path=...` parameter and it is correctly setting the path from the command line argument, no matter what format of path is passed.

The problem resides here in `is_ignored_by_top_level_config()`

<https://github.com/biomejs/biome/blob/main/crates/biome_service/src/workspace/server.rs#L302-L327>

More specifically, the call to `ignore.matched_path_or_any_parents(...)` is throwing an assertion panic because it contains an empty string for the `root` property.

https://github.com/biomejs/biome/blob/main/crates/biome_service/src/workspace/server.rs#L317-L325

So, really, it looks like when nothing or a full absolute path is passed as parameter for `--config-path=...`, a correct absolute path is set for `ignore.root`. But when any non-absolute path is passed as a parameter to `--config-path=...` (eg, `--config-path=biome.jsonc`, `--config-path=.`, etc) `ignore.root == ''`. `to_git_ignore()` called by `to_file_settings()` which is called by `merge_with_configuration()` which is called by `update_settings()`, etc.

Suffice to say, if you set `ignore.root` correctly (by passing a correct base path down to `to_git_ignore()`) when initializing the app settings, that should fix the problem.

As an aside, the error message for a missing file is misleading. You should probably add code to produce an error earlier in load_config() from *crates/biome_service/src/configuration.rs* to produce an understandable error for the user. As it is now, this falls through to testing for `biome.json`, `biome.jsonc`, ... as suffixes to the passed path, even when that path doesn't exist and the tests will always fail. And the resultant error message mentions a path that the user never entered.

Here's something like what I'm talking about...

```rust
    // If the configuration path hint is from user and is a file path,
    // we'll load it directly
    if let ConfigurationPathHint::FromUser(ref config_file_path) = base_path {
        if file_system.path_is_file(config_file_path) {
            let content = file_system.read_file_from_path(config_file_path)?;
            let parser_options = match config_file_path.extension().map(OsStr::as_encoded_bytes) {
                Some(b"json") => JsonParserOptions::default(),
                _ => JsonParserOptions::default()
                    .with_allow_comments()
                    .with_allow_trailing_commas(),
            };
            let deserialized =
                deserialize_from_json_str::<PartialConfiguration>(&content, parser_options, "");
            return Ok(Some(ConfigurationPayload {
                deserialized,
                configuration_file_path: PathBuf::from(config_file_path),
                external_resolution_base_path,
            }));
        }
        // *** ADDED - start
        if !file_system.path_exists(config_file_path) {
            eprintln!("error: the configuration file/path '{config_file_path:?}' doesn't exist");
            return Err(
                crate::WorkspaceError::not_found(),
                // crate::WorkspaceError::not_found().with_file_path(config_file_path),
            );
        }
        // *** ADDED - end
    }
```

**Comment by sambacha** [comment]
On macOS I have observed a similar issue as a result of this shim that is generated at install:

```bash
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")

case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
esac

if [ -z "$NODE_PATH" ]; then
  export NODE_PATH="/Users/guest/reproduce_issue/node_modules/.pnpm/@biomejs+biome@1.9.4/node_modules/@biomejs/biome/bin/node_modules:/Users/guest/reproduce_issue/node_modules/.p>
else
  export NODE_PATH="/Users/guest/reproduce_issue/node_modules/.pnpm/@biomejs+biome@1.9.4/node_modules/@biomejs/biome/bin/node_modules:/Users/guest/reproduce_issue/node_modules/ .p>
fi
if [ -x "$basedir/node" ]; then
  exec "$basedir/node"  "$basedir/../@biomejs/biome/bin/biome" "$@"
else
  exec node  "$basedir/../@biomejs/biome/bin/biome" "$@"
fi


```

Not sure why there is this `node_modules/.p>` and this `node_modules/ .p>` being appended to the `NODE_PATH`, it invalidates the export.

**Comment by ematipico** [comment]
@rivy 

Thank you for the extensive explanation. It's really a weird one, because I really can't replicate it, however I will try again to debug it.

Regarding the configuration file, that should be already covered by `file_system.read_file_from_path(config_file_path)?;`. If the file can't be read (doesn't exist), an error is thrown.
